File: server/docker/sandbox/runner.py
import json
import sys

INPUT_FILE = "/app/input.json"
OUTPUT_FILE = "/app/output.json"

try: 
    with open(INPUT_FILE, "r") as f:
        data = json.load(f)
    
    code = data["code"]
    function_name = data["functionName"]
    test_cases = data["testCases"]

    local_scope = {}

    exec(code, {}, local_scope)

    passed = 0
    results = []

    for test in test_cases:
        args = eval(test["input"])
        expected = json.load(test["expectedOutput"])

        output = local_scope[function_name](*args)

        is_correct = output == expected
        if is_correct:
            passed += 1

        results.append({
            "input": test["input"],
            "expected": expected,
            "output": output,
            "passed": is_correct
        })

    with open(OUTPUT_FILE, "w") as f:
        json.dump({
            "passed": passed,
            "total": len(test_cases),
            "results": results
        }, f)

except Exception as e:
    with open(OUTPUT_FILE, "w") as f:
        json.dump({ "error": str(e) }, f)

File: server/docker/sandbox/runner.js
const fs = require("fs");
const vm = require("vm");

const input = JSON.parse(fs.readFileSync("/app/input.json", "utf-8"));

const { code, functionName, testCases } = input;

let passed = 0;
let results = [];

try {
    const context = {};
    vm.createContext(context);

    const script = new vm.Script(code);
    script.runInContext(context, { timeout: 1000 });

    for (const testCase of testCases) {
        const args = eval(`[${testCase.input}]`);
        const expected = JSON.parse(testCase.expectedOutput);

        const output = context[functionName](...args);


        const isCorrect = JSON.stringify(output) === JSON.stringify(expected);

        if (isCorrect) passed++;

        results.push({
            input: testCase.input,
            expected,
            output,
            passed: isCorrect,
        });
    }

    fs.writeFileSync(
        "/app/output.json",
        JSON.stringify({ passed, total: testCases.length, results })
    );
    
}
catch (err) {
    fs.writeFileSync(
        "/app/output.json",
        JSON.stringify({ error: err.message })
    );
}